File: backend/app/routes/beers.py
from fastapi import APIRouter, HTTPException, Depends
from app.db import get_db
from app.models import CachedBeer
from app.schemas import BeerSearch
from typing import Optional
from pydantic import BaseModel
from bson import ObjectId

router = APIRouter(prefix="/beers", tags=["beers"])

class BeerCreate(BaseModel):
    name: str
    brewery: Optional[str] = None
    type: Optional[str] = None
    abv: Optional[str] = None
    ibu: Optional[str] = None
    description: Optional[str] = None

@router.post("/add")
async def add_beer(beer: BeerCreate, db=Depends(get_db)):
    # Generate a unique _id
    beer_doc = beer.dict()
    beer_doc["_id"] = str(ObjectId())
    # Check for duplicate by name and brewery
    existing = await db.beers.find_one({"name": beer_doc["name"], "brewery": beer_doc.get("brewery")})
    if existing:
        raise HTTPException(status_code=400, detail="Beer already exists")
    await db.beers.insert_one(beer_doc)
    return {"message": "Beer added", "beer": beer_doc}

@router.get("/search")
async def search_beers(
    name: Optional[str] = None,
    brewery: Optional[str] = None,
    type: Optional[str] = None,
    keywords: Optional[str] = None,
    db=Depends(get_db)
):
    query = {}
    if name:
        query["name"] = {"$regex": name, "$options": "i"}
    if brewery:
        query["brewery"] = {"$regex": brewery, "$options": "i"}
    if type:
        query["type"] = {"$regex": type, "$options": "i"}
    if keywords:
        # Search in name, brewery, type, description
        query["$or"] = [
            {"name": {"$regex": keywords, "$options": "i"}},
            {"brewery": {"$regex": keywords, "$options": "i"}},
            {"type": {"$regex": keywords, "$options": "i"}},
            {"description": {"$regex": keywords, "$options": "i"}},
        ]
    beers = await db.beers.find(query).to_list(50)
    return beers

@router.get("/{beer_id}")
async def get_beer(beer_id: str, db=Depends(get_db)):
    beer = await db.beers.find_one({"_id": beer_id})
    if not beer:
        raise HTTPException(status_code=404, detail="Beer not found")
    return beer 

File: backend-ts/src/routes/beers.ts
import express, { Request, Response } from 'express';
import { ObjectId } from 'mongodb';
import { getDatabase } from '../config/database';
import { authenticate, AuthenticatedRequest } from '../utils/auth';
import { BeerCreate, BeerResponse } from '../models/types';

const router = express.Router();

// Search beers
router.get('/search', async (req: Request, res: Response) => {
  try {
    const { q = '', beer_type } = req.query;
    const db = getDatabase();
    
    let query: any = {};
    
    // Add text search if query provided
    if (q && typeof q === 'string' && q.trim()) {
      query.$or = [
        { name: { $regex: q, $options: 'i' } },
        { brewery: { $regex: q, $options: 'i' } }
      ];
    }
    
    // Add beer type filter if provided
    if (beer_type && typeof beer_type === 'string') {
      query.type = beer_type;
    }
    
    const beers = await db.collection('beers')
      .find(query)
      .limit(50)
      .toArray();
    
    const beerResponses: BeerResponse[] = beers.map(beer => ({
      _id: beer._id.toString(),
      name: beer.name,
      brewery: beer.brewery,
      type: beer.type,
      abv: beer.abv,
      ibu: beer.ibu,
      description: beer.description,
      image_url: beer.image_url,
      created_at: beer.created_at
    }));
    
    res.json(beerResponses);
  } catch (error) {
    console.error('Search error:', error);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

// Get beer by ID
router.get('/:id', async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const db = getDatabase();
    
    let beer;
    
    // Try to find beer by ObjectId first (for new beers)
    if (ObjectId.isValid(id)) {
      beer = await db.collection('beers').findOne({ _id: new ObjectId(id) });
    }
    
    // If not found and ID is not ObjectId, try finding by string ID (for legacy beers)
    if (!beer) {
      beer = await db.collection('beers').findOne({ _id: id });
    }
    
    if (!beer) {
      return res.status(404).json({ detail: 'Beer not found' });
    }

    const beerResponse: BeerResponse = {
      _id: beer._id.toString(),
      name: beer.name,
      brewery: beer.brewery,
      type: beer.type,
      abv: beer.abv,
      ibu: beer.ibu,
      description: beer.description,
      image_url: beer.image_url,
      created_at: beer.created_at
    };

    res.json(beerResponse);
  } catch (error) {
    console.error('Get beer error:', error);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

// Add new beer (authenticated)
router.post('/add', authenticate, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const beerData: BeerCreate = req.body;
    
    if (!beerData.name || !beerData.brewery || !beerData.type) {
      return res.status(400).json({ detail: 'Name, brewery, and type are required' });
    }
    
    const db = getDatabase();
    
    // Check if beer already exists
    const existingBeer = await db.collection('beers').findOne({
      name: beerData.name,
      brewery: beerData.brewery
    });
    
    if (existingBeer) {
      return res.status(400).json({ detail: 'Beer already exists' });
    }
    
    const newBeer = {
      ...beerData,
      created_at: new Date()
    };
    
    const result = await db.collection('beers').insertOne(newBeer);
    
    const beerResponse: BeerResponse = {
      _id: result.insertedId.toString(),
      name: newBeer.name,
      brewery: newBeer.brewery,
      type: newBeer.type,
      abv: newBeer.abv,
      ibu: newBeer.ibu,
      description: newBeer.description,
      image_url: newBeer.image_url,
      created_at: newBeer.created_at
    };
    
    res.status(201).json(beerResponse);
  } catch (error) {
    console.error('Add beer error:', error);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

export default router;
